### DISCUSSION TITLE
"This header is not available in freestanding mode."

### DISCUSSION THREAD

**Discussion by NoNamedCat** [py32duino/Arduino-PY32#6] [General]
### 描述一下这个bug / Describe the bug

when I try to compile the simple "blink" program the output says:

.arduino15/packages/PY32Duino/tools/xpack-arm-none-eabi-gcc/13.2.1-1.1/arm-none-eabi/include/c++/13.2.1/bits/requires_hosted.h:34:4: error: #error "This header is not available in freestanding mode."
   34 | #  error "This header is not available in freestanding mode."
   

### 复现步骤 / To Reproduce

v0.1.1
System: Linux Mint

### 如果正常，应该是什么样 / Expected behavior

Compile the sketch.

### 截图 / Screenshots

![image](https://github.com/py32duino/Arduino-PY32/assets/39467683/25600f28-8122-42da-905a-beb237dc32a9)

### 日志 / Logs

.arduino15/packages/PY32Duino/tools/xpack-arm-none-eabi-gcc/13.2.1-1.1/arm-none-eabi/include/c++/13.2.1/bits/requires_hosted.h:34:4: error: #error "This header is not available in freestanding mode."
   34 | #  error "This header is not available in freestanding mode."

### 系统 / System

Mint 22.04

### PACK包版本 / Version

0.1.1

### 验证

- [X] 检查过该问题，之前没有人提过 / Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] 提供了最小可复现工程或详细的复现步骤，确保开发者可以复现 / The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.
- [X] 已经提供了完整的报错信息、日志、截图，没有经过删减。

**Comment by HalfSweet** [comment]
Fixed in 0.1.2

**Comment by NoNamedCat** [comment]
Tks HalfSweet

**Comment by NoNamedCat** [comment]
Sorry. With the 0.1.2 i have this issue:

WARNING: library SrcWrapper claims to run on AirMCU architecture(s) and may be incompatible with your current board which runs on PY32 architecture(s).
In file included from /home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/system/PY32F0xx/py32f0xx_hal_conf_default.h:220,
                 from /home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/system/PY32F0xx/py32f0xx_hal_conf.h:13,
                 from /home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/system/Arduino-PY32F0xx-Drivers/PY32F0xx_HAL_Driver/Inc/py32f0xx_hal.h:33,
                 from /home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/system/Arduino-PY32F0xx-Drivers/CMSIS/Device/PY32F0xx/Include/py32f0xx.h:195,
                 from /home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/cores/arduino/py32/py32_def.h:26,
                 from /home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/cores/arduino/py32/clock.h:19,
                 from /home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/cores/arduino/wiring_time.h:23,
                 from /home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/cores/arduino/wiring.h:38,
                 from /home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/cores/arduino/Arduino.h:36,
                 from /tmp/arduino/sketches/2C45634B9C38D38B552DD500A30E66FE/sketch/sketch_dec1a.ino.cpp:1:
/home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/system/Arduino-PY32F0xx-Drivers/PY32F0xx_HAL_Driver/Inc/py32f0xx_hal_dma.h:116:3: error: 'DMA_Channel_TypeDef' does not name a type; did you mean 'DMA_InitTypeDef'?
  116 |   DMA_Channel_TypeDef   *Instance;                                                    /*!< Register base address                  */
      |   ^~~~~~~~~~~~~~~~~~~
      |   DMA_InitTypeDef
/home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/system/Arduino-PY32F0xx-Drivers/PY32F0xx_HAL_Driver/Inc/py32f0xx_hal_dma.h:136:3: error: 'DMA_TypeDef' does not name a type; did you mean 'TIM_TypeDef'?
  136 |   DMA_TypeDef          *DmaBaseAddress;                                               /*!< DMA Channel Base Address                */
      |   ^~~~~~~~~~~
      |   TIM_TypeDef
In file included from /home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/system/PY32F0xx/py32f0xx_hal_conf_default.h:284:
/home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/system/Arduino-PY32F0xx-Drivers/PY32F0xx_HAL_Driver/Inc/py32f0xx_hal_rtc.h:173:3: error: 'RTC_TypeDef' does not name a type; did you mean 'RCC_TypeDef'?
  173 |   RTC_TypeDef                 *Instance;  /*!< Register base address    */
      |   ^~~~~~~~~~~
      |   RCC_TypeDef
/home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/cores/arduino/py32/backup.h: In function 'void resetBackupDomain()':
/home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/system/Arduino-PY32F0xx-Drivers/PY32F0xx_HAL_Driver/Inc/py32f0xx_hal_rcc.h:1251:60: error: 'RCC_BDCR_BDRST' was not declared in this scope
 1251 | #define __HAL_RCC_BACKUPRESET_FORCE()   SET_BIT(RCC->BDCR, RCC_BDCR_BDRST)
      |                                                            ^~~~~~~~~~~~~~
/home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/system/Arduino-PY32F0xx-Drivers/CMSIS/Device/PY32F0xx/Include/py32f0xx.h:170:42: note: in definition of macro 'SET_BIT'
  170 | #define SET_BIT(REG, BIT)     ((REG) |= (BIT))
      |                                          ^~~
/home/nigrum/.arduino15/packages/PY32Duino/hardware/PY32/0.1.2/cores/arduino/py32/backup.h:71:3: note: in expansion of macro '__HAL_RCC_BACKUPRESET_FORCE'
   71 |   __HAL_RCC_BACKUPRESET_FORCE();
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~

exit status 1

Compilation error: exit status 1

**Comment by HalfSweet** [comment]
What development board model did you choose?

**Comment by NoNamedCat** [comment]
PY32F002

**Comment by NoNamedCat** [comment]
When I choose the PY32F030 i have another problem:

WARNING: library SrcWrapper claims to run on AirMCU architecture(s) and may be incompatible with your current board which runs on PY32 architecture(s).
/home/nigrum/.arduino15/packages/PY32Duino/tools/xpack-arm-none-eabi-gcc/13.2.1-1.1/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: /tmp/arduino/sketches/2C45634B9C38D38B552DD500A30E66FE/sketch_dec1a.ino.elf section `._user_heap_stack' will not fit in region `RAM'
/home/nigrum/.arduino15/packages/PY32Duino/tools/xpack-arm-none-eabi-gcc/13.2.1-1.1/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: region `RAM' overflowed by 304 bytes
collect2: error: ld returned 1 exit status

exit status 1

Compilation error: exit status 1

This is with the blink example.

**Comment by HalfSweet** [comment]
When choosing the F003 or F030 models with 2k RAM, it is possible that the linker will complain about not having enough RAM space. This problem is due to the fact that the framework takes up too much of it... and the main Serial library takes up quite a bit.

We can switch to the `Disabled (no Serial support)` option in the `U(S)ART support` settings, or modify the buffer size by manual injection, which will help solve the problem.

**Comment by HalfSweet** [comment]
In version 0.1.3, PY32F002A should compile properly.

Feel free to contact me if you have any problems with the runtime, and any comments you can make about this project would be greatly appreciated.

**Comment by garudaonekh** [comment]
Yes, it works with PY32F002 now.

**Comment by NoNamedCat** [comment]
I don't want to open a new issue for this question HalfSweet. Forgive me if posting this question here is wrong. I will delete it if so.
Do you have info about the py32f002 and its ram?
The datasheet says something about it, but in reality it has 4kb of ram. I have many boards (I purchased the chips on LCSC) and all of the chips report 4kb of ram instead of the 3kB of static RAM that the datasheet says.
All the people on forums have the same behavior on the chips (Seems that the silicon inside is the same as the PY32F030x6(32KB Flash/4KB RAM), in fact in many forums they said that is possible to use the py32f002 as a py32f030x6 without a problem. Is really confusing to me all of this. If I want to design a product with this IC obviulsy is recommended to follow the datasheet and the specs. But why they use a more complete silicon and sell the chips with a lower price?
Also... the py32f series is a clone of the AirMCU family?

**Comment by HalfSweet** [comment]
It's a hard one to average out, maybe the chip makers will use the same core inside multiple chips with different labeling for their own profit maximization or to reduce inventory, which might lead us to misunderstandings about the overall product.

But in any case, this repository will follow the datasheet and you can choose another model if you need to, in theory this will not be a problem.

Both AirMCU and PY32Duino are maintained by me, including OpenPuya in the future, but these are two different companies and eventually there may be some features that will not be fully synchronized.

Finally, thanks for your love of this project, and if you run into any problems, feel free to ask!

**Comment by NoNamedCat** [comment]
I found another company making the same chip? The datasheet is very similar with the specs of the IC. I have to translate and read it. But for me sound like it is the same.
https://www.lcsc.com/product-detail/Microcontroller-Units-MCUs-MPUs-SOCs_Geehy-APM32F003F6P6_C2975251.html

**Comment by NoNamedCat** [comment]
Here is the schematic for a board that I'm designing. What do you think? Are you interested on the design?
![PY-CAT-002](https://github.com/py32duino/Arduino-PY32/assets/39467683/9d7f05cf-fa3e-4a13-bd49-e6de24cf575a)

**Comment by HalfSweet** [comment]
Sorry, I'm not familiar with Geehy as a company.

But it's not surprising for a Cortex-M core processor to have similar or even the same peripheral IP, so if it really doesn't differ much from the py32, maybe I can add it to this project.

**Comment by NoNamedCat** [comment]
I grabbed the bootloader part from your design.
Here is the render of the board:
![bottom](https://github.com/py32duino/Arduino-PY32/assets/39467683/57578281-15fb-4642-b494-fc0dee47fa2f)
![top](https://github.com/py32duino/Arduino-PY32/assets/39467683/f36cc6f9-4de4-44c2-bea3-8076ac07f784)

**Comment by HalfSweet** [comment]
> Here is the schematic for a board that I'm designing. What do you think? Are you interested on the design? ![PY-CAT-002](https://private-user-images.githubusercontent.com/39467683/287611411-9d7f05cf-fa3e-4a13-bd49-e6de24cf575a.svg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE2NzI1MzUsIm5iZiI6MTcwMTY3MjIzNSwicGF0aCI6Ii8zOTQ2NzY4My8yODc2MTE0MTEtOWQ3ZjA1Y2YtZmEzZS00YTEzLWJkNDktZTZkZTI0Y2Y1NzVhLnN2Zz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA0VDA2NDM1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyNWE0ZTEwMGUzMGYxMmMxNzFmYTBhYTgwYzE2MDQzMTM4ZDI0MzBhYjU1OGM3OGVjOWM5ODJkMjhkODI2MDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JYwIJL2IQ67TcQx9h57jLdcppMaRj8xTxXgtghkCIb8)

One thing worth noting is that the PY32 doesn't necessarily interact well with 3.3V level peripherals - either UARTs or SWDs - when it's running on a 5V supply, which I think is kind of a pitfall.

And, the V3 pin of the CH340 seems to require an external 0.1uF capacitor, I'm not quite sure, you'll need to check the datasheet

**Comment by HalfSweet** [comment]
It's probably not a good idea to have chat in issues, I've turned it into discussions where we can have

**Reply by NoNamedCat** [reply]
Sorry. My fault. Not happening again.

**Comment by NoNamedCat** [comment]
The capacitor is needed in the 3.3v output pin for voltage stability when you use 5v as VCC. If you feed all the circuit with an stable 3.3v you don't need that cap.
Also in my tests with the old design for the same chip that also can switch the voltage (3.3v or 5v) all the flashing work flawlesly.

**Comment by HalfSweet** [comment]
Moreover, the CH340X supports the so-called "microcontroller USB one-click download", so if it works, maybe you can skip the three BJT designs.

Actually, the CH343P also supports a similar feature, plus it's just to make the design more versatile...

If you intend to continue with such a design, you should note that NRST requires a 10k pull-up resistor and a 1uF to ground capacitor; BOOT0 requires both a 10k and a 1uF to ground capacitor.

**Comment by HalfSweet** [comment]
![图片](https://github.com/py32duino/Arduino-PY32/assets/60973476/891c7272-e85c-472c-bcd0-eeefd43df2d9)

**Comment by NoNamedCat** [comment]
Tks a lot HalfSweet. You are very kind for giving me that advice. I propose that on another repo. Thats why im using the CH340X on the design. But that serial IC is not working as intended. For the test that another person do it doesnt work well. Can i show you the repo? He has a collection of designs for many IC and also have a repo of programs to flash many MCU's. I think that you will like it.

**Reply by HalfSweet** [reply]
Sure. I thought we could have a friendly chat.

**Reply by NoNamedCat** [reply]
Here is the repo for the flash tools:
https://github.com/wagiminator/MCU-Flash-Tools

And here you can find the templates for various microcontrollers.
https://github.com/wagiminator/MCU-Templates

In the repo for the flash tools you can find the problem with the CH340x:
https://github.com/wagiminator/MCU-Flash-Tools/issues/2

**Reply by NoNamedCat** [reply]
I was designing this board for that MCU before you posted the nano sized board (is awesome). I believe that with the production and sale of that board, this project will attract more people. It's incredible to me how they managed to reduce the cost of that microcontroller so much!
